File: uni_modules/ui-component/utssdk/app-ios/swift/observer/BarrageStoreObserver.swift
import Combine
import DCloudUTSFoundation
import RTCRoomEngine_Plus

public class BarrageStoreObserver {
    private var cancellables = Set<AnyCancellable>()
    public static let shared = BarrageStoreObserver()
    
    public func barrageStoreChanged(_ liveId: String, _ callback: @escaping (_ name: String, _ data: String) -> Void) {
        cancellables.removeAll()

        V2RTCRoomEngine.shared.getBarrageStore(liveId: liveId)
            .state.subscribe(StateSelector(keyPath: \V2BarrageState.messageList))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] messageList in
                guard let self = self else { return }
                var list = []
                for messageInfo in messageList {
                    if (messageInfo != nil) {
                        list.append(convertMessageInfoToDic(messageInfo: messageInfo))
                    }
                }
                if let jsonList = JsonUtil.toJson(list) {
                    callback("messageList", jsonList)
                }
            })
            .store(in: &cancellables)
    }

    private func convertMessageInfoToDic(messageInfo: MessageInfo) -> [String: Any] {
        var dict: [String: Any] = [
            "roomId": messageInfo.roomId,
            "sender": TypeConvert.convertTUIUserInfoToDic(userInfo: messageInfo.sender),
            "sequence": messageInfo.sequence,
            "timestampInSecond": messageInfo.timestampInSecond,
            "messageType": convertMessageType(messageInfo.messageType),
            "textContent": messageInfo.textContent,
            "extensionInfo": messageInfo.extensionInfo,
            "businessId": messageInfo.businessId,
            "data": messageInfo.data
        ]
        return dict
    }

    private func convertMessageType(_ type: MessageType) -> String {
        if (type == MessageType.custom) {
            return "CUSTOM" 
        }
        return "TEXT"
    }

    
} 

File: uni_modules/ui-component/utssdk/app-android/kotlin/observer/BarrageStoreObserver.kt
package uts.sdk.modules.uiComponent.observer

import com.google.gson.Gson
import com.tencent.cloud.uikit.state.MessageInfo
import com.tencent.cloud.uikit.state.V2RTCRoomEngine
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import uts.sdk.modules.uiComponent.kotlin.observer.TypeConvert

object BarrageStoreObserver {
    private val gson = Gson()
    private var bindDataJob: Job? = null

    fun barrageStoreChanged(liveId: String, callback: (String, String) -> Unit) {
        bindDataJob?.cancel()
        bindDataJob = CoroutineScope(Dispatchers.Main).launch {
            launch {
                V2RTCRoomEngine.getInstance().getBarrageStore(liveId).barrageState.messageList.collect { messageList ->
                    val list = messageList.map { convertMessageInfoToMap(it) }
                    callback("messageList", gson.toJson(list))
                }
            }
        }
    }

    private fun convertMessageInfoToMap(messageInfo: MessageInfo): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        map["roomId"] = messageInfo.roomId
        map["sender"] = TypeConvert.convertUserInfoToMap(messageInfo.sender)
        map["sequence"] = messageInfo.sequence
        map["timestampInSecond"] = messageInfo.timestampInSecond
        map["messageType"] = convertMessageType(messageInfo.messageType.value)
        map["textContent"] = messageInfo.textContent
        map["extensionInfo"] = messageInfo.extensionInfo
        map["businessId"] = messageInfo.businessId
        map["data"] = messageInfo.data
        return map
    }

    private fun convertMessageType(type: Int): String {
        if (type == 1) {
            return "CUSTOM"
        }
        return "TEXT"
    }
}
